logstash-core/src/main/java/org/logstash/instrument/metrics/HistoricMetric.java

```diff
@@ -0,0 +1,230 @@
+ * This directionality allows us to maintain an efficient insertion-time cache of nodes by age and
+ * to compact expired entries without full iteration.
+ * It is tolerant to a reasonable magnitude of out-of-order insertions.
+ *
+ * At insertion time, we maintain links to useful nodes, rotating them forward as needed to avoid full-scale
+ * iteration in either read- or write-modes. These nodes are defined as the _youngest_ node satisfying their
+ * category's minimum time threshold, and their accuracy is dependent on regularly-scheduled _captures_ of the
+ * underlying metrics.
+ */
+public class HistoricMetric<T extends Number> {
+
+    // Test Dependency Injection
+    private final Clock clock;
+
+    // Metric Sources
+    private final Metric<T> numeratorMetric;
+    private final Metric<Long> denominatorMetric;
+
+    // Oldest and Newest nodes in linked list. Oldest *usable* capture is head.next
+    private final Capture head = new Capture(null, null, null);
+    private final AtomicReference<Capture> tail = new AtomicReference<>(head);
+
+    // Important intermediate nodes in our linked list
+    private final QuickReference quickInstant = new QuickReference("current", Duration.ofSeconds(1));
+    private final QuickReference quickOneMinute = new QuickReference("last_1_minute", Duration.ofMinutes(1));
+    private final QuickReference quickFiveMinutes = new QuickReference("last_5_minutes", Duration.ofMinutes(5));
+    private final QuickReference quickFifteenMinutes = new QuickReference("last_15_minutes", Duration.ofMinutes(15));
+    private final QuickReference quickLifetime = new QuickLifetimeReference("lifetime");
+
+    // update- and retrieve-order quick reference lists.
+    private final List<QuickReference> quickUpdateReferences = List.of(quickLifetime, quickInstant, quickOneMinute, quickFiveMinutes, quickFifteenMinutes);
+    private final List<QuickReference> quickReadReferences =   List.of(quickInstant, quickOneMinute, quickFiveMinutes, quickFifteenMinutes, quickLifetime);
+
+    // Safeguard for reads against a dead writer
+    private final AtomicReference<Instant> lastCaptured = new AtomicReference<>();
+    private static final Duration STALENESS_THRESHOLD = Duration.ofSeconds(10);
+
+    public HistoricMetric(Metric<T> numeratorMetric, Metric<Long> denominatorMetric) {
+        this(Clock.systemUTC(), numeratorMetric, denominatorMetric);
+    }
+
+    public HistoricMetric(Metric<T> metric) {
+        this(Clock.systemUTC(), metric);
+    }
+
+    HistoricMetric(Clock clock, Metric<T> numeratorMetric, Metric<Long> denominatorMetric) {
+        this.clock = Objects.requireNonNull(clock, "clock must not be null");
+        this.numeratorMetric = Objects.requireNonNull(numeratorMetric, "numeratorMetric must not be null");
+        this.denominatorMetric = Objects.requireNonNullElseGet(denominatorMetric, () -> new UptimeMetric(clock, null));
+    }
```

Comment:
```suggestion
    HistoricMetric(@Nonnull Clock clock, @Nonnull Metric<T> numeratorMetric, Optional<Metric<Long>> denominatorMetric) {
        this.clock = clock;
        this.numeratorMetric = numeratorMetric;
        this.denominatorMetric = denominatorMetric.orElse(new UptimeMetric(clock, null));
    }
```

If suggestion LGTY, we shouldn't forget to `import javax.annotation.Nonnull;`